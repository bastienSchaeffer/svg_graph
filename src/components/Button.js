// @flow

import React from 'react';

type ButtonProps = {
  children: string,
};

const Button = ({ children }: ButtonProps) => (
  <div>
    <button type="button">{children}</button>
  </div>
);

export default Button;
