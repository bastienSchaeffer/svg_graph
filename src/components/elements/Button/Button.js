// @flow

import React, { type Node } from 'react';
import StyledButton from './Button.styles';

type ButtonProps = {
  children: Node,
  className?: string,
  disabled?: boolean,
  fullWidth?: boolean,
  loading?: boolean,
  onClick?: MouseEvent => void | Promise<void>,
  palette?: 'primary' | 'secondary' | 'secondarySolid' | 'naked',
  testId?: string,
  type?: 'button' | 'submit',
  size?: 'sm' | 'md' | 'lg',
  circle?: boolean,
};

const Button = ({
  children,
  className,
  disabled,
  fullWidth,
  loading = false,
  onClick,
  palette = 'primary',
  size = 'lg',
  testId,
  type = 'button',
  circle = false,
}: ButtonProps) => (
  <StyledButton
    className={className}
    data-test-id={testId}
    disabled={disabled || loading}
    fullWidth={fullWidth}
    loading={loading}
    onClick={onClick}
    palette={palette}
    size={size}
    type={type}
    circle={circle}
  >
    {children}
  </StyledButton>
);

export default Button;
