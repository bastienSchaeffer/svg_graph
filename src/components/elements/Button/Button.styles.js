import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';
import { ifProp, switchProp } from 'styled-tools';
import {
  // Colors
  brightText,
  darkText,
  elevation,
  primaryColor,
  primaryColorDark,
  secondaryColor,
  secondaryColorDark,
  // Fonts
  fontSize2,
  fontSize3,
  fontSize4,
  bold,
  // Media
  fromSmall,
  // Shape
  circle,
} from '../../../styles';

/*
  Property: type
*/
const primary = css`
  background: ${primaryColor};
  color: ${brightText};
  &:hover {
    background: ${primaryColorDark};
  }
`;

const secondary = css`
  background: ${secondaryColor};
  color: ${brightText};
  &:hover {
    background: ${secondaryColorDark};
  }
`;

const naked = css`
  background: none;
  color: ${darkText};
`;

/*
  Property: size
*/
const sm = css`
  ${fontSize4};
  font-weight: 500;
  padding: 4px 8px;
  /* ${fromSmall(css`
    font-size: 44px;
    line-height: 54px;
  `)}; */
  ${ifProp(
    'circle',
    css`
      ${circle('45px')}
    `,
  )};
`;

const md = css`
  ${fontSize3};
  font-weight: 500;
  padding: 6px 16px;
  ${ifProp(
    'circle',
    css`
      ${circle('50px')}
    `,
  )};
`;
const lg = css`
  ${fontSize2};
  font-weight: 500;
  padding: 8px 24px;
  ${ifProp(
    'circle',
    css`
      ${circle('60px')}
    `,
  )};
`;

/*
  StyledButton
*/
const StyledButton = styled.button`
  appearance: none;
  outline: 0;
  text-transform: uppercase;
  border-radius: 4px;
  display: inline-block;
  background-clip: padding-box;
  transition: background-color 0.3s ease-out;
  ${elevation[1]};
  border: 0;
  cursor: pointer;
  margin: 0 8px 8px 0;
  ${switchProp('palette', { primary, secondary, naked })};
  ${switchProp('size', { sm, md, lg })};
  ${ifProp(
    'fullWidth',
    css`
      width: 100%;
      padding-left: 0;
      padding-right: 0;
      margin-left: 0;
      margin-right: 0;
    `,
  )};

  &:disabled {
    cursor: default;
    box-shadow: none;
    background-color: #e0e0e0;
    color: ${darkText};
  }
`;

StyledButton.propTypes = {
  disabled: PropTypes.bool,
  palette: PropTypes.oneOf(['primary', 'secondary', 'naked']),
  size: PropTypes.oneOf(['lg', 'md', 'sm']),
};

StyledButton.defaultProps = {
  disabled: false,
  size: 'lg',
};

export default StyledButton;
