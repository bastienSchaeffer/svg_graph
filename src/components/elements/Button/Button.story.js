import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { withKnobs, text, select } from '@storybook/addon-knobs';
import { withInfo } from '@storybook/addon-info';

import Button from './Button';

/*
  Dropdown knob Size
*/
const label = 'Size';
const sizeOptions = {
  Small: 'sm',
  Medium: 'md',
  large: 'lg',
};
const defaultValue = 'sm';

/*
  Button stories
*/
storiesOf('Elements', module)
  .addDecorator(withKnobs)
  .addDecorator(withInfo({ header: false, inline: true, source: false }))
  .add('Button', () => (
    <div>
      <h1>Button</h1>
      <hr />
      <h2>Size</h2>
      <Button size="sm">Button: sm</Button>
      <Button size="md">Button: md</Button>
      <Button size="lg">Button: lg</Button>
      <hr />
      <h2>Type</h2>
      <Button palette="primary">Button: primary</Button>
      <Button palette="secondary">Button: secondary</Button>
      <Button palette="naked">Button: naked</Button>
      <hr />
      <h2>Other:</h2>
      <Button disabled={true}>Button: :disabled</Button>
      <Button fullWidth={true}>Button: fullWidth</Button>
      <hr />
    </div>
  ));

// size={select(label, sizeOptions, defaultValue)}
// {text('Label', 'Button')}
