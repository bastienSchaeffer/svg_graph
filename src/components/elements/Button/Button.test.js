import React from 'react';
import { shallow } from 'enzyme';

import Button from './Button';

describe('Button:', () => {
  let component;

  const buttonProps = {
    size: 'sm',
    onClick: jest.fn(),
  };

  beforeEach(() => {
    component = shallow(<Button {...buttonProps}>Button</Button>);
  });

  describe('Render', () => {
    it('should render with minimum config', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('Behaviour', () => {
    it('should handle the click event passed', () => {
      buttonProps.onClick.mockClear();
      component.simulate('click');
      expect(buttonProps.onClick).toHaveBeenCalledTimes(1);
    });
  });
});
