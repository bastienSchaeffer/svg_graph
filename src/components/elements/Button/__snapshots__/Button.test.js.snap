// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button: Render should render with minimum config 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Button
    onClick={[MockFunction]}
    size="sm"
  >
    Button
  </Button>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": "Button",
      "className": undefined,
      "data-test-id": undefined,
      "disabled": false,
      "fullWidth": undefined,
      "loading": false,
      "onClick": [MockFunction],
      "palette": "primary",
      "size": "sm",
      "type": "button",
    },
    "ref": null,
    "rendered": "Button",
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "attrs": Array [],
      "componentStyle": ComponentStyle {
        "componentId": "sc-bdVaJa",
        "isStatic": false,
        "rules": Array [
          "
  appearance: none;
  outline: 0;
  display: inline-block;
  background-clip: padding-box;
  transition: box-shadow 0.3s ease-out, background-color 0.3s ease-out;
  cursor: pointer;
  margin: 0;
  ",
          [Function],
          ";
  ",
          [Function],
          ";
  ",
          [Function],
          ";

  &:disabled {
    cursor: default;
    box-shadow: none;
    background-color: red;
    color: green;
    border-color: purple;
  }

  &:hover {
    ",
          "box-shadow: 0 1px 3px rgba(0,0,0,.12), 0 1px 2px rgba(0,0,0,.24)",
          "
  }
",
        ],
      },
      "defaultProps": Object {
        "disabled": false,
        "size": "lg",
      },
      "displayName": "styled.button",
      "foldedComponentIds": Array [],
      "propTypes": Object {
        "disabled": [Function],
        "palette": [Function],
        "size": [Function],
      },
      "render": [Function],
      "styledComponentId": "sc-bdVaJa",
      "target": "button",
      "toString": [Function],
      "warnTooManyClasses": [Function],
      "withComponent": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": "Button",
        "className": undefined,
        "data-test-id": undefined,
        "disabled": false,
        "fullWidth": undefined,
        "loading": false,
        "onClick": [MockFunction],
        "palette": "primary",
        "size": "sm",
        "type": "button",
      },
      "ref": null,
      "rendered": "Button",
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "sc-bdVaJa",
          "isStatic": false,
          "rules": Array [
            "
  appearance: none;
  outline: 0;
  display: inline-block;
  background-clip: padding-box;
  transition: box-shadow 0.3s ease-out, background-color 0.3s ease-out;
  cursor: pointer;
  margin: 0;
  ",
            [Function],
            ";
  ",
            [Function],
            ";
  ",
            [Function],
            ";

  &:disabled {
    cursor: default;
    box-shadow: none;
    background-color: red;
    color: green;
    border-color: purple;
  }

  &:hover {
    ",
            "box-shadow: 0 1px 3px rgba(0,0,0,.12), 0 1px 2px rgba(0,0,0,.24)",
            "
  }
",
          ],
        },
        "defaultProps": Object {
          "disabled": false,
          "size": "lg",
        },
        "displayName": "styled.button",
        "foldedComponentIds": Array [],
        "propTypes": Object {
          "disabled": [Function],
          "palette": [Function],
          "size": [Function],
        },
        "render": [Function],
        "styledComponentId": "sc-bdVaJa",
        "target": "button",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
